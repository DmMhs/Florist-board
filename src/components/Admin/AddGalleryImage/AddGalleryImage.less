.gallery-info {
  text-align: left;
  width: 100%;
}
