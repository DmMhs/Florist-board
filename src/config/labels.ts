const labels = {
  brand: 'FLORIST.UA',
  adress: 'MARKET SQUARE, LVIV, UA',
  postCode: '346346',
  phone: '+123-45-678',
  navigation: {
    home: 'HOME',
    shop: 'SHOP',
    gallery: 'GALLERY',
    contacts: 'CONTACTS',
    cart: 'CART',
    account: {
      main: 'MY ACCOUNT',
      menu: {
        signIn: 'SIGN IN',
        signUp: 'SIGN UP'
      }
    }
  }
};

export default labels;
