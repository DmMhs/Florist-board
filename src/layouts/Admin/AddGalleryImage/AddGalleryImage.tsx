import React, { Component } from 'react';

import { AppContext } from '../../../AppContext';
import { storageRef, galleryImagesRef } from '../../../firebase';
import { uploadGalleryImage } from '../../../services/admin/uploadGalleryImage';
import './AddGalleryImage.less';
import { getGalleryImageDownloadURL } from '../../../services/admin/getGalleryImageDownloadURL';

interface AddGalleryImageProps {}

interface AddGalleryImageState {
  images: File[];
  totalImagesNumber: number;
}

class AddGalleryImage extends Component<
  AddGalleryImageProps,
  AddGalleryImageState
> {
  constructor(props: AddGalleryImageProps) {
    super(props);
    this.state = {
      images: [],
      totalImagesNumber: 0
    };
  }

  public componentDidMount() {
    galleryImagesRef.on('value', snapshot => {
      this.setState({
        totalImagesNumber: Object.keys(snapshot!.val()).length
      });
    });
  }

  private imgInputChangedHandler = (
    position: number,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedImages: File[] = [...this.state.images];
    updatedImages[position] = event.target.files as any;
    this.setState({
      images: updatedImages
    });
  };

  private formSubmitHandler = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const { images } = this.state;
    this.setState({
      images: []
    });

    images.map(async (image: File) => {
      const file = (image as any)[0];
      const formattedFileName = (image as any)[0].name;
      await uploadGalleryImage(file, formattedFileName);
      getGalleryImageDownloadURL(formattedFileName)
        .then(res => {
          return res;
        })
        .then(imgURL => {
          galleryImagesRef.push(imgURL);
        });
    });
  };

  public render() {
    const context = this.context;
    const labels = context.state.labels;
    const lang = context.state.lang;

    const labelsRoot = labels[lang].pages.admin.addGalleryImageForm;
    const submitBtnLabel = labels[lang].pages.admin.submitBtn;
    const color = this.state.totalImagesNumber % 4 === 0 ? 'green' : 'darkred';
    return (
      <AppContext.Consumer>
        {value =>
          value && (
            <form
              onSubmit={this.formSubmitHandler}
              className="AddGalleryImage form"
            >
              <div className="form-control product-images">
                <p style={{ color }} className="gallery-info">
                  {labelsRoot.info.restriction}
                </p>
                <p className="gallery-info">
                  {labelsRoot.info.totalImagesNumber}{' '}
                  {this.state.totalImagesNumber}
                </p>
                <hr />
                <label>{labelsRoot.add}</label>
                <br />
                <div className="input-wrapper">
                  <input
                    type="file"
                    onChange={this.imgInputChangedHandler.bind(this, 0)}
                  />
                </div>
              </div>
              <button type="submit">{submitBtnLabel}</button>
            </form>
          )
        }
      </AppContext.Consumer>
    );
  }
}

AddGalleryImage.contextType = AppContext;

export default AddGalleryImage;
