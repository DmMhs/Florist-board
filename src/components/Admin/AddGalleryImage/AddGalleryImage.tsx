import React, { Component } from 'react';

import './AddGalleryImage.less';
import { AppContext } from '../../../AppContext';
import { storageRef, galleryImagesRef } from '../../../firebase';
import labels from '../../../config/labels';

interface AddGalleryImageProps {}

interface AddGalleryImageState {
  images: File[];
  totalImagesNumber: number;
}

class AddGalleryImage extends Component<
  AddGalleryImageProps,
  AddGalleryImageState
> {
  constructor(props: AddGalleryImageProps) {
    super(props);
    this.state = {
      images: [],
      totalImagesNumber: 0
    };
  }

  public componentDidMount() {
    galleryImagesRef.on('value', snapshot => {
      this.setState({
        totalImagesNumber: Object.keys(snapshot!.val()).length
      });
    });
  }

  private imgInputChangedHandler = (
    position: number,
    event: React.ChangeEvent<HTMLInputElement>
  ) => {
    const updatedImages: File[] = [...this.state.images];
    updatedImages[position] = event.target.files as any;
    this.setState({
      images: updatedImages
    });
  };

  private formSubmitHandler = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const { images } = this.state;
    this.setState({
      images: []
    });

    images.map(async (image: File) => {
      const file = (image as any)[0];
      const formattedFileName = (image as any)[0].name;
      await storageRef
        .child('gallery-images')
        .child(formattedFileName)
        .put(file)
        .catch(err => {
          console.log(err);
        });
      const imageURL = await storageRef
        .child('gallery-images')
        .child(formattedFileName)
        .getDownloadURL();
      galleryImagesRef.push(imageURL);
    });
  };

  public render() {
    const context = this.context;
    const labelsRoot =
      labels[context.state.lang as string].pages.admin.addGalleryImageForm;
    const submitBtnLabel =
      labels[context.state.lang as string].pages.admin.submitBtn;
    const color = this.state.totalImagesNumber % 4 === 0 ? 'green' : 'darkred';
    return (
      <AppContext.Consumer>
        {value =>
          value && (
            <form onSubmit={this.formSubmitHandler} className="AddProduct form">
              <div className="form-control product-images">
                <p style={{ color }} className="gallery-info">
                  {labelsRoot.info.restriction}
                </p>
                <p className="gallery-info">
                  {labelsRoot.info.totalImagesNumber}{' '}
                  {this.state.totalImagesNumber}
                </p>
                <hr />
                <label>{labelsRoot.add}</label>
                <br />
                <div className="input-wrapper">
                  <input
                    type="file"
                    onChange={this.imgInputChangedHandler.bind(this, 0)}
                  />
                </div>
              </div>
              <button type="submit">{submitBtnLabel}</button>
            </form>
          )
        }
      </AppContext.Consumer>
    );
  }
}

AddGalleryImage.contextType = AppContext;

export default AddGalleryImage;
