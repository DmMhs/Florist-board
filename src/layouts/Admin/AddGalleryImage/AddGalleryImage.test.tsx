import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';

import AddGalleryImage from './AddGalleryImage';
import AppContextProvider from '../../../AppContext';
import labels from '../../../config/labels';

describe('AddGalleryImage works as expected', () => {
  it('AddGalleryImage component matches a snapshot', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddGalleryImage />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
  it('total images number is not 0', () => {
    const wrapper = mount(
      <BrowserRouter>
        <AppContextProvider>
          <AddGalleryImage />
        </AppContextProvider>
      </BrowserRouter>
    );
    const context = wrapper.find('AppContextProvider').instance();
    context.setState({
      lang: 'en',
      labels: labels,
      fetchInProgress: false,
      mobileMode: true,
      showNavigation: false,
      togglePosition: 'absolute'
    });
    wrapper.update();
    const instance = wrapper.find('AddGalleryImage').instance();
    instance.setState({
      images: ['one', 'two', 'three']
    });
    wrapper.find('.form').simulate('submit');
    wrapper.update();
    expect(instance.state).toEqual({
      images: [],
      totalImagesNumber: 0
    });
  });
});
