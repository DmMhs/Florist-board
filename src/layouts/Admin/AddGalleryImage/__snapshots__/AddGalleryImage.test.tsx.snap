// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddGalleryImage works as expected AddGalleryImage component matches a snapshot 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <withRouter(AppContextProvider)>
      <Route>
        <AppContextProvider
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
        >
          <AddGalleryImage>
            <form
              className="AddGalleryImage form"
              onSubmit={[Function]}
            >
              <div
                className="form-control product-images"
              >
                <p
                  className="gallery-info"
                  style={
                    Object {
                      "color": "green",
                    }
                  }
                >
                  Expression "(Number of images)/4" should has no remainder for the best gallery displaying
                </p>
                <p
                  className="gallery-info"
                >
                  Number of images in gallery: 
                   
                  0
                </p>
                <hr />
                <label>
                  Add image to the galery:
                </label>
                <br />
                <div
                  className="input-wrapper"
                >
                  <input
                    onChange={[Function]}
                    type="file"
                  />
                </div>
              </div>
              <button
                type="submit"
              >
                SUBMIT
              </button>
            </form>
          </AddGalleryImage>
        </AppContextProvider>
      </Route>
    </withRouter(AppContextProvider)>
  </Router>
</BrowserRouter>
`;
